src/shared/tpm2-util.c

```diff
@@ -4127,6 +4127,30 @@ int tpm2_parse_luks2_json(
         return 0;
 }
 
+size_t tpm2_hash_alg_to_size(uint16_t alg) {
+        if (alg == TPM2_ALG_SHA1)
+                return 20;
+        if (alg == TPM2_ALG_SHA256)
+                return 32;
+        if (alg == TPM2_ALG_SHA384)
+                return 48;
+        if (alg == TPM2_ALG_SHA512)
+                return 64;
+        return 0;
+}
+
+int tpm2_hash_alg_from_size(size_t size) {
```

Comment:
uh, this smells like it's going to break sooner or later, no? if another hash alg is added with the same size?

Reply:
quite true, and that's [already the case](https://en.wikipedia.org/wiki/SHA-2#Comparison_of_SHA_functions) for sha3 family of hashes (which have the same size as sha2 hashes). I'll rework this.

I added this in order to avoid users needing to specify the hash to --tpm2-pcrs, since all the sha1/sha2 algs have different hash sizes. Do you think we should always require the hash alg, or should we keep the 'detection' of the right hash but only for the sha1/sha2 algs?

for example should we allow this:
`systemd-cryptenroll --tpm2-pcrs 1=0x0000000000000000000000000000000000000000000000000000000000000000`
which would 'detect' that the hash is sha256 from the hash length

while if users wanted to use sha3-256, they could specify the alg (technically this won't work now, since we don't parse for sha3 yet):
`systemd-cryptenroll --tpm2-pcrs 1:sha3-256=0x0000000000000000000000000000000000000000000000000000000000000000`

Making the hash alg optional is strictly for the convenience of users, so if you think we should require it i have no objections.

Reply:
i pushed a new version with `tpm2_hash_alg_from_size()` removed, and `tpm2_parse_pcr_digest_spec()` doing the SHA1/SHA2 'detection' based on hash value length. I can update to remove that and require specifying the hash alg if you think that's better, let me know.

Reply:
after looking at it again, i removed the 'optional' part - a hash algorithm  has to be provided now when specifying a hash digest value. the 'convenience' of omitting the hash alg isn't really needed.

Reply:
i removed the CLEANUP_ERASE() in favor of TAKE_STRUCT() and re-pushed...figured we dont need to do both.